File: test/crc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


unsigned int iriCRC(unsigned char *buf, int cnt) {
  long accum=0x00000000;
  int i, j;
  if (cnt <= 0) return 0;
  while (cnt--) {
    accum |= *buf++ & 0xFF;
    for (i = 0; i < 8; i++) {
      accum <<= 1;
      if (accum & 0x01000000)
        accum ^= 0x00102100;
    }
    printf("accum %x\n", accum);
  }
  // compatibility with XMODEM CRC
  for (j = 0; j < 2; j++) {
    // accum |= 0 & 0xFF;
    for (i = 0; i < 8; i++) {
      accum <<= 1;
      if (accum & 0x01000000)
        accum ^= 0x00102100;
    }
    printf("accum %x\n", accum);
  }
  return (accum >> 8 & 0xFFFF);
} // iriCRC

int main() {
  char *buff;
  char *strings[3];
  unsigned int cs;
  FILE *fd;
  int cnt, len, i;
  //
  buff=malloc(64);
  // platform
  len = 8;
  sprintf(buff, "???CS%4s%4s", "LR01", "QUEH");
  cs = iriCRC(buff+5, len);
  buff[3] = (char) (cs >> 8) & 0xFF;
  buff[4] = (char) (cs & 0xFF);
  printf("\n crc %x %u \n", cs, cs);
  fd = fopen("platform.hdr", "w");
  fwrite(buff, sizeof(char), len+5, fd);
  fclose(fd);
  // message
  len = 10;
  sprintf(buff, "@@@CS%c%cT%c%c%5s", 
      (char) (len>>8 & 0xFF), (char) (len & 0xFF),
      1, 1, "Hello");
  cs = iriCRC(buff+5, len);
  buff[3] = (char) (cs >> 8) & 0xFF;
  buff[4] = (char) (cs & 0xFF);
  printf("\n crc %x %u \n", cs, cs);
  fd = fopen("message.blk", "w");
  fwrite(buff, sizeof(char), len+5, fd);
  fclose(fd);
  // some strings
  printf(" some strings\n");
  strings[0]="Medium size";
  strings[1]="longish asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf";
  strings[2]="long lllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll";
  for (i=0; i<1; i++) {
    cs = iriCRC(strings[i], strlen(strings[i]));
    printf("str %d cs %x %u\n", i, cs, cs);
  }
}

File: test/crc.py
#!/usr/bin/python
#
import io
#
def sum(str):
  """rudics 16bit CRC (bytes str)"""
  accum=0
  mask=0x7FFFFFFF
  for b in str:
    accum |= b
    # python int will get huge - strip bit 16 (a&7F)<<1
    for i in range(0,8):
      accum = (accum & mask) << 1
      if (accum & 0x01000000):
        accum ^= 0x00102100
  # xmodem dumb
  for i in range(0,2):
    # accum |= 0
    for i in range(0,8):
      accum = (accum & mask) << 1
      if (accum & 0x01000000):
        accum ^= 0x00102100
  return  ((accum >> 8) & 0xFFFF)
#
def msg(str):
  """return blk:bytes for (str:bytes)"""
  l=len(str)+5
  b=l.to_bytes(2,'big')+b'T\x01\x01'+str
  cs=sum(b)
  blk=b'@@@'+cs.to_bytes(2,'big')+b
  return blk
#
h=b'LR01'+b'QUEH'
cs=sum(h)
hdr=b'???'+cs.to_bytes(2,'big')+h
blk=msg(b'Hello')
f=open("py.hdr","wb")
f.write(hdr)
f.close()
f=open("py.blk","wb")
f.write(blk)
f.close()


